library/std/math.qs

```diff
@@ -499,7 +665,288 @@ namespace Microsoft.Quantum.Math {
+        Fact(a >= 0L, "`a` must be non-negative.");
+        mutable number = a;
+        mutable size = 0;
+        while (number != 0L) {
+            set size = size + 1;
+            set number = number >>> 1;
+        }
+
+        size
+    }
+
+    //
+    // Combinatorics
+    //
+
+    /// # Summary
+    /// Returns the factorial of a given number.
+    ///
+    /// # Description
+    /// Returns the factorial of a given nonnegative integer n, where 0 ≤ n ≤ 20.
+    ///
+    /// # Input
+    /// ## n
+    /// The number to take the factorial of.
+    ///
+    /// # Output
+    /// The factorial of `n`.
+    ///
+    /// # Remarks
+    /// For inputs greater than 20, please use `Microsoft.Quantum.Math.FactorialL`.
+    ///
+    /// # See Also
+    /// - Microsoft.Quantum.Math.FactorialL
+    /// - Microsoft.Quantum.Math.ApproximateFactorial
+    function FactorialI(n : Int) : Int {
+        Fact(n >= 0, "The factorial is not defined for negative inputs.");
+        Fact(n <= 20, "The largest factorial that be stored as an Int is 20!. Use FactorialL or ApproximateFactorial.");
+
+        mutable result = 1;
+        for i in 1 .. n {
+            set result *= i;
+        }
+        result
+    }
+
+    /// # Summary
+    /// Returns the factorial of a given number.
+    ///
+    /// # Description
+    /// Returns the factorial of a given nonnegative integer n, where 0 ≤ n ≤ 20.
+    ///
+    /// # Input
+    /// ## n
+    /// The number to take the factorial of.
+    ///
+    /// # Output
+    /// The factorial of `n`.
+    ///
+    /// # See Also
+    /// - Microsoft.Quantum.Math.FactorialI
+    /// - Microsoft.Quantum.Math.ApproximateFactorial
+    function FactorialL(n : Int) : BigInt {
+        Fact(n >= 0, "The factorial is not defined for negative inputs.");
+
+        mutable result = 1L;
+        for i in 1 .. n {
+            set result *= IntAsBigInt(i);
+        }
+        result
+    }
+
+    /// # Summary
+    /// Returns an approximate factorial of a given number.
+    ///
+    /// # Description
+    /// Returns the factorial as `Double`, given an input `n`.
+    /// The domain of inputs for this function is `n <= 169`.
+    ///
+    /// # Remarks
+    /// For n > 10, this function uses the Ramanujan approximation with a
+    /// relative error of the order of 1 / n⁵.
+    ///
+    /// # Input
+    /// ## n
+    /// The number to take the approximate factorial of. Must not be negative.
+    ///
+    /// # Output
+    /// The approximate factorial of `n`.
+    ///
+    /// # See Also
+    /// - Microsoft.Quantum.Math.FactorialI
+    /// - Microsoft.Quantum.Math.FactorialL
+    function ApproximateFactorial(n : Int) : Double {
+        Fact(n >= 0, "The factorial is not defined for negative inputs.");
+        Fact(n <= 169, "The largest approximate factorial that be stored as an Double is 169!. Use FactorialL.");
+
+        // For small enough n, use the exact factorial instead.
+        if n < 10 {
```

Comment:
Why not 20 if that's what FactorialI supports?

Reply:
I updated FactorialI implementation and changed this function to use it when possible.
I think that the point was to use fewer operations if the result can be approximate. I changed factorial to use precomputed array of 21 element so that the execution time is constant. Now this function can use FactorialI whenever possible.